internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubUtil.java

```diff
@@ -178,4 +180,75 @@ public static boolean isPubSubSslProtocol(String pubSubSecurityProtocol) {
       return false; // or rethrow if desired
     }
   }
+
+  /**
+   * Computes the difference in numeric offsets between two {@link PubSubPosition} instances.
+   *
+   * <p>If both positions are {@code EARLIEST} or both are {@code LATEST}, returns {@code 0}.</p>
+   * <p>If one of the positions is {@code EARLIEST} and the other is {@code LATEST},
+   * it returns {@code Long.MIN_VALUE} or {@code Long.MAX_VALUE} to represent symbolic extremes.</p>
+   *
+   * @param position1 The first position.
+   * @param position2 The second position.
+   * @return The numeric difference: (position1 - position2).
+   * @throws IllegalArgumentException if either position is {@code null}.
+   */
+  public static long diffPubSubPositions(PubSubPosition position1, PubSubPosition position2) {
+    if (position1 == null || position2 == null) {
+      throw new IllegalArgumentException("Positions cannot be null");
+    }
+
+    if ((position1 == PubSubSymbolicPosition.EARLIEST && position2 == PubSubSymbolicPosition.EARLIEST)
+        || (position1 == PubSubSymbolicPosition.LATEST && position2 == PubSubSymbolicPosition.LATEST)) {
+      return 0;
+    }
+
+    if (position1 == PubSubSymbolicPosition.EARLIEST && position2 == PubSubSymbolicPosition.LATEST) {
+      return Long.MIN_VALUE;
+    }
+
+    if (position1 == PubSubSymbolicPosition.LATEST && position2 == PubSubSymbolicPosition.EARLIEST) {
+      return Long.MAX_VALUE;
+    }
+
+    return position1.getNumericOffset() - position2.getNumericOffset();
+  }
+
+  /**
+   * Compares two {@link PubSubPosition} instances by their symbolic or numeric ordering.
+   *
+   * <p>This method defines the following ordering:
+   * EARLIEST &lt; numeric offsets &lt; LATEST</p>
+   *
+   * @param position1 The first position.
+   * @param position2 The second position.
+   * @return A negative integer, zero, or a positive integer if position1 is less than,
+   *         equal to, or greater than position2, respectively.
+   * @throws IllegalArgumentException if either position is {@code null}.
+   */
+  public static int comparePubSubPositions(PubSubPosition position1, PubSubPosition position2) {
+    if (position1 == null || position2 == null) {
+      throw new IllegalArgumentException("Positions cannot be null");
+    }
+
+    if ((position1 == PubSubSymbolicPosition.EARLIEST && position2 == PubSubSymbolicPosition.EARLIEST)
+        || (position1 == PubSubSymbolicPosition.LATEST && position2 == PubSubSymbolicPosition.LATEST)) {
+      return 0;
+    }
+
+    if (position1 == PubSubSymbolicPosition.EARLIEST) {
+      return -1;
+    }
+    if (position1 == PubSubSymbolicPosition.LATEST) {
+      return 1;
+    }
+    if (position2 == PubSubSymbolicPosition.EARLIEST) {
+      return 1;
+    }
+    if (position2 == PubSubSymbolicPosition.LATEST) {
+      return -1;
+    }
+
+    return Long.compare(position1.getNumericOffset(), position2.getNumericOffset());
```

Comment:
should we update getNumericOffset() of EARLIEST and LATEST to return MIN_VALUE and MAX_VALUE ?
it is error prone and inconvenient for calls to check for these two special type of offsets before every comparison

Reply:
I’d prefer not to. The goal here is to ensure implementations handle symbolic positions directly, as numeric offsets are being phased out.